### PR TITLE
End to end testing

### PR DESCRIPTION
**Pull request by finestructure**
@daveverwer These are a few small fixes that cropped up while end-to-end testing processing all whole 2.7k packages.

I’ve now successfully processed everything locally and on my Digital Ocean box *) except for 5 packages I had to delete manually because they brought up the git login prompt (see #27 ) -EDIT: the commit I just added fixes this. It went well all in all!

I’ll send you the db details for the Digital Ocean DB to have a look at the data. We’ve got close to 30k versions and around 26k products (plus the 2.7 packages and repositories) in there now.

Have a look through the data - it should be mostly complete, except for a few minor omissions, I think (default branch versions and fork links, most notably). The main purpose was to run it end to end once to make sure processing isn’t interrupted by errors bubbling up and to have the full table structure with all entities in place.

*) The DO box ran out of disk space with 100 packages left to process 😅 - so that data is almost complete as well. The checkout size is 16.6GB now.

### CODE REVIEW

File: Sources/App/Commands/Analyzer.swift [hunk 1]
```diff
@@ -57,18 +81,27 @@ func analyze(application: Application, limit: Int) throws -> EventLoopFuture<Voi
 
     // FIXME: sas 2020-05-04: Workaround for partial flush described here:
     // https://discordapp.com/channels/431917998102675485/444249946808647699/706796431540748372
+    // FIXME: this breaks exposing this via the API:
+    // Precondition failed: BUG DETECTED: wait() must not be called when on an EventLoop.
```
Comment (finestructure) [inline Sources/App/Commands/Analyzer.swift:85]: I’ll add an issue to track this, maybe someone can jump in to give a hand.

File: Sources/App/Commands/Analyzer.swift [hunk 2]
```diff
@@ -161,6 +201,11 @@ func getManifest(package: Package, version: Version) -> Result<Manifest, Error>
             throw AppError.invalidRevision(version.id, version.tagName)
         }
         try Current.shell.run(command: .gitCheckout(branch: revision), at: cacheDir)
+        guard Current.fileManager.fileExists(atPath: cacheDir + "/Package.swift") else {
+            // It's important to check for Package.swift - otherwise `dump-package` will go
+            // up the tree through parent directories to find one
+            throw AppError.invalidRevision(version.id, "no Package.swift")
+        }
```
Comment (finestructure) [inline Sources/App/Commands/Analyzer.swift:208]: This whole `analyze` function needs some refactoring and cleanup but I don’t want to do that before we’ve fixed the `wait` workaround.

File: Sources/App/Commands/Analyzer.swift [hunk 3]
```diff
@@ -57,18 +81,27 @@ func analyze(application: Application, limit: Int) throws -> EventLoopFuture<Voi
...
+                .flatMapEach(on: application.db.eventLoop) { result -> EventLoopFuture<Void> in
+                    switch result {
+                        case .success:
+                            pkg.status = .ok
+                        case .failure(let error):
+                            application.logger.error("Analysis error: \(error.localizedDescription)")
+                            pkg.status = .analysisFailed
+                    }
+                    return pkg.save(on: application.db)
+            }
```
Comment (finestructure) [inline Sources/App/Commands/Analyzer.swift:104]: This whole `analyze` function needs some refactoring and cleanup but I don’t want to do that before we’ve fixed the `wait` workaround.

File: Sources/App/Commands/Analyzer.swift (hunk 1 above)
Comment (daveverwer) [reply Sources/App/Commands/Analyzer.swift:85]: I'm not sure I can be much help here, but let's raise an issue for it so it doesn't get forgotten after this is merged.

Comment (finestructure) [reply Sources/App/Commands/Analyzer.swift:85]: Ah yes, thanks for the reminder to actually paste that in :D

#29

**Pull request merged**